engine/src/main/java/org/terasology/rendering/nui/layers/ingame/metrics/DebugMetricsSystem.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.nui.layers.ingame.metrics;
+
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.registry.Share;
+
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+
+/**
+ * Provides MetricsMode objects to DebugOverlay. Default metrics nodes can be modified via register
+ * unregister and unregisterAll functions.
+ */
+@RegisterSystem
+@Share(DebugMetricsSystem.class)
+public class DebugMetricsSystem extends BaseComponentSystem {
+
+    private LinkedHashSet<MetricsMode> metricsModes;
+    private Iterator<MetricsMode> modeIterator;
+    private MetricsMode currentMode;
+
+    @Override
+    public void initialise() {
+        metricsModes = Sets.newLinkedHashSet();
+        modeIterator = Iterators.cycle(metricsModes);
```

Comment:
Please double check all methods that modify the LinkedHashSet. From what I read the iterator might need to be regenerated every time something is added or removed from the set. If that is the case it might be appropriate to have additional register(ArrayList) and unregister(ArrayList) methods to pool changes together and regenerate the iterator only once.